File: discovery-scripts/discover-sources.js
const fetch = require('node-fetch');
const { JSDOM } = require('jsdom');
const fs = require('fs');
const path = require('path');

// RSS feeds to process
const rssFeeds = [
  'https://greenpointers.com/feed',
  'https://www.6sqft.com/feed',
  'https://newyorkyimby.com/feed',
  'https://www.brooklynpaper.com/feed',
  'https://brooklyneagle.com/feed',
  'https://www.bkmag.com/feed',
  'https://canarsiecourier.com/feed',
  'https://northbrooklynnews.com/feed',
  'https://bushwickdaily.com/feed',
  'https://www.star-revue.com/feed',
  'https://brownstoner.com/feed',
  'https://www.amny.com/feed',
  'https://www.thecity.nyc/feed',
  'https://hellgatenyc.com/all-posts/rss/',
  'https://gothamist.com/feed',
  'https://www.boropark24.com/feed',
  'https://politicsny.com/feed/',
  'https://www.caribbeanlifenews.com/feed',
  'https://www.brooklynreporter.com/feed'
];

// Track discovered sources to avoid duplicates
const discoveredSources = new Set();
let sourceCount = 0;
const maxSources = 1000;

function extractExternalLinks(html, feedDomain) {
  const dom = new JSDOM(html);
  const links = dom.window.document.querySelectorAll('a');
  const externalLinks = new Set();
  
  links.forEach(link => {
    const href = link.href;
    try {
      const url = new URL(href);
      if (url.hostname !== feedDomain && !href.startsWith('#') && !href.startsWith('mailto:')) {
        externalLinks.add(href);
      }
    } catch (e) {}
  });
  
  return Array.from(externalLinks);
}

function categorizeSource(url, content = '') {
  const urlLower = url.toLowerCase();
  const contentLower = content.toLowerCase();
  
  // Government sources
  if (url.includes('.gov') || url.includes('nyc.gov') || url.includes('cityrecord')) {
    return 'government_website';
  }
  
  // NYC Open Data
  if (url.includes('opendata.nyc.gov') || url.includes('data.cityofnewyork.us')) {
    return 'nyc_open_data';
  }
  
  // Social media
  if (url.includes('facebook.com') || url.includes('instagram.com')) {
    return 'social_media';
  }
  if (url.includes('twitter.com') || url.includes('x.com')) {
    return 'social_media_twitter';
  }
  if (url.includes('linkedin.com')) {
    return 'social_media_linkedin';
  }
  
  // Real estate
  if (url.includes('streeteasy') || url.includes('zillow') || url.includes('realtor.com') || 
      url.includes('compass.com') || url.includes('corcoran')) {
    return 'real_estate_website';
  }
  
  // Business directories
  if (url.includes('yellowpages') || url.includes('yelp') || url.includes('maps.google')) {
    return 'business_directory';
  }
  
  // Schools/Education
  if (url.includes('.edu') || url.includes('schools.nyc.gov') || contentLower.includes('school')) {
    return 'educational_institution';
  }
  
  // News/Media
  if (url.includes('nytimes') || url.includes('nj.com') || url.includes('amny.com') ||
      url.includes('gothamist') || url.includes('brownstoner')) {
    return 'news_website';
  }
  
  // Community organizations
  if (contentLower.includes('community board') || contentLower.includes('nonprofit') || 
      contentLower.includes('ngo') || url.includes('.org')) {
    return 'community_organization';
  }
  
  // Default
  return 'website';
}

function determineAccessType(sourceType, url) {
  if (url.includes('opendata.nyc.gov') || url.includes('api.')) {
    return 'api_request';
  }
  if (url.includes('.gov')) {
    return 'website_page_scrape';
  }
  if (sourceType === 'social_media_twitter') {
    return 'twitter_api';
  }
  if (sourceType === 'social_media') {
    return 'facebook_graph_api';
  }
  return 'website_page_scrape';
}

function calculateValueScore(sourceType, url, content = '') {
  let score = 5; // Base score
  
  // Government sources are highly valuable
  if (sourceType === 'government_website') score += 3;
  if (sourceType === 'nyc_open_data') score += 4;
  
  // NYC-specific data is more relevant
  if (url.includes('nyc.gov') || url.includes('brooklyn')) score += 1;
  
  // Cap at 10
  return Math.min(score, 10);
}

function createSourceObject(url, feedUrl) {
  const sourceType = categorizeSource(url);
  const accessType = determineAccessType(sourceType, url);
  
  return {
    source_type: sourceType,
    source_access_type: accessType,
    source_client_key: {
      how_to_access: `Access via ${accessType}. URL: ${url}\nDiscovered from feed: ${feedUrl}`
    },
    value_score: calculateValueScore(sourceType, url),
    discovered_from: feedUrl
  };
}

async function fetchRssFeed(feedUrl) {
  try {
    const response = await fetch(feedUrl);
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return await response.text();
  } catch (error) {
    console.error(`Failed to fetch ${feedUrl}:`, error.message);
    return null;
  }
}

async function processFeed(feedUrl) {
  console.log(`\nProcessing: ${feedUrl}`);
  
  const rssContent = await fetchRssFeed(feedUrl);
  if (!rssContent) return 0;
  
  // Parse RSS to extract article content and links
  const dom = new JSDOM(rssContent);
  const items = dom.window.document.querySelectorAll('item');
  
  let sourcesFound = 0;
  const feedDomain = new URL(feedUrl).hostname;
  
  for (const item of items) {
    if (sourceCount >= maxSources) break;
    
    // Get article content
    const description = item.querySelector('description')?.textContent || '';
    const link = item.querySelector('link')?.textContent || '';
    
    // Extract external links from description
    const externalLinks = extractExternalLinks(description, feedDomain);
    
    for (const extLink of externalLinks) {
      if (sourceCount >= maxSources) break;
      
      // Skip duplicates
      if (discoveredSources.has(extLink)) continue;
      
      discoveredSources.add(extLink);
      const source = createSourceObject(extLink, feedUrl);
      
      // Save source to file
      await saveSource(source);
      sourcesFound++;
      sourceCount++;
    }
  }
  
  console.log(`Found ${sourcesFound} new sources from this feed`);
  return sourcesFound;
}

async function saveSource(source) {
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
  const sourceName = `${source.source_type}_${timestamp}`;
  const filename = `${sourceName}.jsonl`;
  
  fs.writeFileSync(filename, JSON.stringify(source) + '\n');
  
  // Commit the file
  await execAsync(`git add ${filename}`);
  await execAsync(`git commit -m "fix: create ${sourceName} source"`);
  
  console.log(`Saved and committed: ${filename}`);
}

function execAsync(command) {
  return new Promise((resolve, reject) => {
    const { exec } = require('child_process');
    exec(command, (error, stdout, stderr) => {
      if (error) reject(error);
      else resolve(stdout);
    });
  });
}

async function main() {
  console.log(`Starting source discovery for Kensington, Brooklyn (11218)`);
  console.log(`Target: ${maxSources} sources`);
  
  for (const feed of rssFeeds) {
    if (sourceCount >= maxSources) break;
    await processFeed(feed);
  }
  
  console.log(`\nTotal sources discovered: ${sourceCount}`);
}

main().catch(console.error);

File: discovery-scripts/discover_sources.py
#!/usr/bin/env python3
import requests
import xml.etree.ElementTree as ET
import json
import subprocess
from datetime import datetime
from urllib.parse import urlparse
import os

# RSS feeds to process
RSS_FEEDS = [
    'https://greenpointers.com/feed',
    'https://www.6sqft.com/feed',
    'https://newyorkyimby.com/feed',
    'https://www.brooklynpaper.com/feed',
    'https://brooklyneagle.com/feed',
    'https://www.bkmag.com/feed',
    'https://canarsiecourier.com/feed',
    'https://northbrooklynnews.com/feed',
    'https://bushwickdaily.com/feed',
    'https://www.star-revue.com/feed',
    'https://brownstoner.com/feed',
    'https://www.amny.com/feed',
    'https://www.thecity.nyc/feed',
    'https://hellgatenyc.com/all-posts/rss/',
    'https://gothamist.com/feed',
    'https://www.boropark24.com/feed',
    'https://politicsny.com/feed/',
    'https://www.caribbeanlifenews.com/feed',
    'https://www.brooklynreporter.com/feed'
]

# Track discovered sources to avoid duplicates
discovered_sources = set()
source_count = 0
max_sources = 1000

def categorize_source(url):
    """Categorize a source based on its URL."""
    url_lower = url.lower()
    
    # Government sources
    if '.gov' in url or 'nyc.gov' in url:
        return 'government_website'
    
    # NYC Open Data
    if 'opendata.nyc.gov' in url or 'data.cityofnewyork.us' in url:
        return 'nyc_open_data'
    
    # Social media - Facebook/Instagram
    if 'facebook.com' in url or 'instagram.com' in url:
        return 'social_media'
    
    # Twitter/X
    if 'twitter.com' in url or 'x.com' in url:
        return 'social_media_twitter'
    
    # LinkedIn
    if 'linkedin.com' in url:
        return 'social_media_linkedin'
    
    # Real estate
    if any(x in url for x in ['streeteasy', 'zillow', 'realtor.com', 'compass.com']):
        return 'real_estate_website'
    
    # Business directories
    if any(x in url for x in ['yellowpages', 'yelp', 'maps.google']):
        return 'business_directory'
    
    # Schools/Education
    if '.edu' in url or 'schools.nyc.gov' in url:
        return 'educational_institution'
    
    # News/Media
    if any(x in url for x in ['nytimes.com', 'nj.com', 'amny.com']):
        return 'news_website'
    
    # Community organizations
    if '.org' in url:
        return 'community_organization'
    
    return 'website'

def determine_access_type(source_type, url):
    """Determine how to access the source."""
    if 'opendata.nyc.gov' in url or 'api.' in url:
        return 'api_request'
    if source_type == 'social_media_twitter':
        return 'twitter_api'
    if source_type == 'social_media':
        return 'facebook_graph_api'
    return 'website_page_scrape'

def calculate_value_score(source_type, url):
    """Calculate a value score from 1-10."""
    score = 5  # Base score
    
    # Government sources are highly valuable
    if source_type == 'government_website':
        score += 3
    if source_type == 'nyc_open_data':
        score += 4
    
    # NYC-specific data is more relevant
    if 'nyc.gov' in url or 'brooklyn' in url:
        score += 1
    
    return min(score, 10)

def extract_links_from_text(text):
    """Extract all URLs from text content."""
    import re
    urls = re.findall(r'https?://[^\s<>"]+', text)
    return [url.rstrip('/') for url in urls]

def parse_rss_feed(feed_url):
    """Parse an RSS feed and extract article content."""
    try:
        response = requests.get(feed_url, timeout=30)
        response.raise_for_status()
        
        # Parse XML
        root = ET.fromstring(response.content)
        
        # Find all items (articles)
        items = root.findall('.//item')
        
        return items, urlparse(feed_url).netloc
    except Exception as e:
        print(f"  Error parsing {feed_url}: {e}")
        return [], None

def save_source(source_data):
    """Save a single source to a file and commit it."""
    global source_count
    
    timestamp = datetime.utcnow().strftime('%Y-%m-%dT%H-%M-%S')
    source_type = source_data['source_type']
    source_name = f"{source_type}_{timestamp}"
    filename = f"{source_name}.jsonl"
    
    # Write to file
    with open(filename, 'w') as f:
        json.dump(source_data, f)
        f.write('\n')
    
    # Commit the file
    try:
        subprocess.run(['git', 'add', filename], check=True, capture_output=True)
        subprocess.run(['git', 'commit', '-m', f"fix: create {source_name} source"], 
                      check=True, capture_output=True)
    except subprocess.CalledProcessError as e:
        print(f"  Git error: {e}")
    
    print(f"  Saved: {filename} (type: {source_type}, score: {source_data['value_score']})")
    source_count += 1

def process_feed(feed_url):
    """Process a single RSS feed."""
    global source_count, discovered_sources
    
    print(f"\nProcessing: {feed_url}")
    
    if source_count >= max_sources:
        return
    
    items, feed_domain = parse_rss_feed(feed_url)
    
    if not items:
        print("  No items found or error parsing")
        return
    
    sources_found = 0
    
    for item in items:
        if source_count >= max_sources:
            break
        
        # Extract content from item
        description = item.find('description')
        content = description.text if description is not None else ''
        
        # Also check for encoded content
        encoded_content = item.find('{http://purl.org/rss/1.0/modules/content/}encoded')
        if encoded_content is not None:
            content += ' ' + encoded_content.text
        
        # Extract links from content
        links = extract_links_from_text(content)
        
        for link in links:
            if source_count >= max_sources:
                break
            
            # Skip if already discovered
            if link in discovered_sources:
                continue
            
            # Skip internal links (same domain as feed)
            try:
                link_domain = urlparse(link).netloc
                if link_domain == feed_domain:
                    continue
            except:
                continue
            
            # Mark as discovered
            discovered_sources.add(link)
            
            # Create source data
            source_type = categorize_source(link)
            access_type = determine_access_type(source_type, link)
            value_score = calculate_value_score(source_type, link)
            
            source_data = {
                'source_type': source_type,
                'source_access_type': access_type,
                'source_client_key': {
                    'how_to_access': f"Access via {access_type}. URL: {link}\nDiscovered from feed: {feed_url}"
                },
                'value_score': value_score
            }
            
            save_source(source_data)
            sources_found += 1
    
    print(f"  Found {sources_found} new sources from this feed")
    return sources_found

def main():
    """Main function to process all feeds."""
    global source_count
    
    os.chdir('/Users/al/.openclaw/workspace/kensington-sources')
    
    print(f"Starting source discovery for Kensington, Brooklyn (11218)")
    print(f"Target: {max_sources} sources")
    print()
    
    total_found = 0
    for feed in RSS_FEEDS:
        if source_count >= max_sources:
            break
        found = process_feed(feed)
        total_found += found
    
    print(f"\n{'='*60}")
    print(f"Total sources discovered: {source_count}")
    print(f"Target reached: {'YES' if source_count >= max_sources else 'NO'}")

if __name__ == '__main__':
    main()
